File: modules/agent-pinning/ios/AgentPinningModule.swift
import ExpoModulesCore

/**
 Verifies the agent by its public key for one request, rather than by a chain.

 The agent signs its own certificate, so ordinary evaluation rejects it before
 anything else is consulted. The pin replaces the chain as the identity check:
 what is compared is the SHA-256 of the certificate's SubjectPublicKeyInfo, the
 same value the WebSocket policy compares and the same one the agent prints in
 its pairing QR.

 A nil pin is a deliberate trust-on-first-use pairing. The certificate is still
 accepted — there is nothing to check it against yet — and the caller reports
 the pairing as unverified rather than as a pairing.
 */
private final class PinnedSessionDelegate: NSObject, URLSessionDelegate {
  private let expectedPin: String?

  init(expectedPin: String?) {
    self.expectedPin = expectedPin
  }

  func urlSession(
    _ session: URLSession,
    didReceive challenge: URLAuthenticationChallenge,
    completionHandler: @escaping (URLSession.AuthChallengeDisposition, URLCredential?) -> Void
  ) {
    guard challenge.protectionSpace.authenticationMethod == NSURLAuthenticationMethodServerTrust,
          let trust = challenge.protectionSpace.serverTrust else {
      completionHandler(.performDefaultHandling, nil)
      return
    }

    guard let expectedPin else {
      completionHandler(.useCredential, URLCredential(trust: trust))
      return
    }

    guard let presented = AgentPinning.pin(forServerTrust: trust), presented == expectedPin else {
      // Cancelling surfaces to the caller as a failed request, which is what a
      // key that is not the agent's should look like.
      completionHandler(.cancelAuthenticationChallenge, nil)
      return
    }

    completionHandler(.useCredential, URLCredential(trust: trust))
  }
}

public class AgentPinningModule: Module {
  public func definition() -> ModuleDefinition {
    Name("AgentPinning")

    Property("isSupported") { true }

    Function("setPin") { (pin: String?) in
      AgentPinning.setPin(pin)
    }

    // Pairing is a one-off POST that has to be verified by the same pin the
    // socket is, and it happens before any pin is stored — so it cannot go
    // through setPin, and React Native's fetch has no per-request trust hook.
    // This builds a session for this one request and invalidates it after.
    AsyncFunction("postPinned") {
      (url: String, pin: String?, body: String, promise: Promise) in
      guard let endpoint = URL(string: url) else {
        promise.reject("ERR_AGENT_PINNING_URL", "Not a URL: \(url)")
        return
      }

      var request = URLRequest(url: endpoint)
      request.httpMethod = "POST"
      request.setValue("application/json", forHTTPHeaderField: "Content-Type")
      request.httpBody = body.data(using: .utf8)
      request.timeoutInterval = 15

      let configuration = URLSessionConfiguration.ephemeral
      let session = URLSession(
        configuration: configuration,
        delegate: PinnedSessionDelegate(expectedPin: pin),
        delegateQueue: nil
      )

      session.dataTask(with: request) { data, response, error in
        // Without this the session holds its delegate, and with it this closure,
        // for the life of the process.
        session.finishTasksAndInvalidate()

        if let error {
          promise.reject("ERR_AGENT_PINNING_REQUEST", error.localizedDescription)
          return
        }
        guard let http = response as? HTTPURLResponse else {
          promise.reject("ERR_AGENT_PINNING_REQUEST", "The agent's answer was not HTTP.")
          return
        }
        promise.resolve([
          "status": http.statusCode,
          "body": String(data: data ?? Data(), encoding: .utf8) ?? "",
        ])
      }.resume()
    }
  }
}

File: modules/agent-pinning/android/src/main/java/expo/modules/agentpinning/AgentPinningModule.kt
package expo.modules.agentpinning

import android.util.Base64
import com.facebook.react.modules.network.CustomClientBuilder
import com.facebook.react.modules.websocket.WebSocketModule
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import java.security.MessageDigest
import java.security.SecureRandom
import java.security.cert.CertificateException
import java.security.cert.X509Certificate
import java.util.concurrent.TimeUnit
import javax.net.ssl.SSLContext
import javax.net.ssl.X509TrustManager

/**
 * Accepts the agent's certificate when its public key matches the pin handed
 * over at pairing, and refuses it otherwise.
 *
 * This is a trust manager rather than an OkHttp CertificatePinner on purpose.
 * CertificatePinner runs after the chain has been validated, so a self-signed
 * certificate is rejected during the handshake and the pin is never consulted —
 * documented OkHttp behaviour, and the most common way this goes wrong.
 *
 * The key is pinned rather than the certificate because the agent reissues its
 * certificate whenever its host changes network, while keeping the key. Pinning
 * the certificate would break on every reissue.
 */
internal class PinnedTrustManager(private val expectedPin: String) : X509TrustManager {
  override fun checkServerTrusted(chain: Array<X509Certificate>, authType: String) {
    val leaf = chain.firstOrNull() ?: throw CertificateException("agent presented no certificate")

    // PublicKey.getEncoded() is already SPKI DER, which is what the agent hashes.
    val digest = MessageDigest.getInstance("SHA-256").digest(leaf.publicKey.encoded)
    val pin = "sha256/" + Base64.encodeToString(digest, Base64.NO_WRAP)

    if (pin != expectedPin) {
      throw CertificateException("agent key pin mismatch")
    }
  }

  // The agent never asks the device for a certificate.
  override fun checkClientTrusted(chain: Array<X509Certificate>, authType: String) = Unit

  override fun getAcceptedIssuers(): Array<X509Certificate> = emptyArray()
}

internal class PinningClientBuilder(private val pin: String) : CustomClientBuilder {
  override fun apply(builder: OkHttpClient.Builder) {
    val trustManager = PinnedTrustManager(pin)
    val sslContext = SSLContext.getInstance("TLS")
    sslContext.init(null, arrayOf<X509TrustManager>(trustManager), SecureRandom())

    builder.sslSocketFactory(sslContext.socketFactory, trustManager)

    // Hostname verification is deliberately left at the default. The agent's
    // certificate carries its hostnames and LAN addresses as SANs, so ordinary
    // verification passes for any address it is reached at, and a mismatch is
    // worth surfacing rather than suppressing.
  }
}

/**
 * Accepts whatever the agent presents, for a pairing that has no pin yet.
 *
 * Only reachable when the caller asked for an unpinned pairing: the PIN is what
 * authorizes the exchange in that case, and the app reports the result as
 * unverified rather than as a pairing.
 */
internal class AcceptAnyTrustManager : X509TrustManager {
  override fun checkServerTrusted(chain: Array<X509Certificate>, authType: String) = Unit
  override fun checkClientTrusted(chain: Array<X509Certificate>, authType: String) = Unit
  override fun getAcceptedIssuers(): Array<X509Certificate> = emptyArray()
}

class AgentPinningModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("AgentPinning")

    Property("isSupported") { true }

    // React Native's WebSocket does not go through OkHttpClientProvider — its
    // module builds an OkHttpClient of its own — so the client customization
    // that covers fetch does not reach a WebSocket. setCustomClientBuilder is
    // the hook that does.
    Function("setPin") { pin: String? ->
      WebSocketModule.setCustomClientBuilder(pin?.let { PinningClientBuilder(it) })
    }

    // Pairing is a one-off POST that has to be verified by the same pin the
    // socket is, and it happens before any pin is stored — so it cannot go
    // through setPin, and RN's fetch has no per-request trust hook. This builds
    // a client for this one request and throws it away.
    AsyncFunction("postPinned") { url: String, pin: String?, body: String ->
      val builder = OkHttpClient.Builder()
        .connectTimeout(TIMEOUT_SECONDS, TimeUnit.SECONDS)
        .readTimeout(TIMEOUT_SECONDS, TimeUnit.SECONDS)

      // A null pin is a deliberate trust-on-first-use pairing, which still has
      // to reach an agent serving a certificate nothing here signed.
      val trustManager = if (pin != null) PinnedTrustManager(pin) else AcceptAnyTrustManager()
      val sslContext = SSLContext.getInstance("TLS")
      sslContext.init(null, arrayOf<X509TrustManager>(trustManager), SecureRandom())
      builder.sslSocketFactory(sslContext.socketFactory, trustManager)

      // The agent's certificate names its hostnames and LAN addresses, but a
      // device reaching it by an address the certificate does not carry is a
      // working setup: the key pin is the identity check here, and hostname
      // verification would refuse before it is consulted.
      builder.hostnameVerifier { _, _ -> true }

      val request = Request.Builder()
        .url(url)
        .post(body.toRequestBody(JSON_MEDIA_TYPE))
        .build()

      builder.build().newCall(request).execute().use { response ->
        mapOf(
          "status" to response.code,
          "body" to (response.body?.string() ?: ""),
        )
      }
    }
  }

  private companion object {
    const val TIMEOUT_SECONDS = 15L
    val JSON_MEDIA_TYPE = "application/json".toMediaType()
  }
}
